libraries/lib-graphics/graphics/Rect.h

```diff
@@ -0,0 +1,267 @@
+   AxisAlignedBoundingBoxType(AxisAlignedBoundingBoxType &&) = default;
+   AxisAlignedBoundingBoxType &operator=(const AxisAlignedBoundingBoxType &) = default;
+   AxisAlignedBoundingBoxType &operator=(AxisAlignedBoundingBoxType &&) = default;
+
+   //! Expands the box to include the specified point.
+   AxisAlignedBoundingBoxType& Expand(PointType<DataType> point) noexcept
+   {
+      topLeft.x = std::min(topLeft.x, point.x);
+      topLeft.y = std::min(topLeft.y, point.y);
+      bottomRight.x = std::max(bottomRight.x, point.x);
+      bottomRight.y = std::max(bottomRight.y, point.y);
+
+      return *this;
+   }
+
+   //! Expands the copy of the box to include the specified rectangle.
+   AxisAlignedBoundingBoxType Expanded(PointType<DataType> point) const noexcept
+   {
+      AxisAlignedBoundingBoxType result = *this;
+      result.Expand(point);
+      return result;
+   }
+
+   //! Expands the box to include the specified rectangle.
+   AxisAlignedBoundingBoxType& Expand(const RectType<DataType>& rect) noexcept
+   {
+      topLeft.x = std::min(topLeft.x, rect.origin.x);
+      topLeft.y = std::min(topLeft.y, rect.origin.y);
+      bottomRight.x = std::max(bottomRight.x, rect.origin.x + rect.size.width);
+      bottomRight.y = std::max(bottomRight.y, rect.origin.y + rect.size.height);
+
+      return *this;
+   }
+
+   //! Expands the copy of the box to include the specified rectangle.
+   AxisAlignedBoundingBoxType Expanded(const RectType<DataType>& rect) const noexcept
+   {
+      AxisAlignedBoundingBoxType result = *this;
+      result.Expand(rect);
+      return result;
+   }
+
+   //! Expands the box to include the specified box.
+   AxisAlignedBoundingBoxType& Expand(const AxisAlignedBoundingBoxType<DataType>& other) noexcept
+   {
+      topLeft.x = std::min(topLeft.x, other.topLeft.x);
+      topLeft.y = std::min(topLeft.y, other.topLeft.y);
+      bottomRight.x = std::max(bottomRight.x, other.bottomRight.x);
+      bottomRight.y = std::max(bottomRight.y, other.bottomRight.y);
+
+      return *this;
+   }
+
+   //! Expands the copy of the box to include the specified box.
+   AxisAlignedBoundingBoxType Expanded(const AxisAlignedBoundingBoxType other) const noexcept
+   {
+      AxisAlignedBoundingBoxType result = *this;
+      result.Expand(other);
+      return result;
+   }
+
+   //! Returns true if the box contains the specified point.
+   bool Contains(PointType<DataType> point) const noexcept
+   {
+      return (point.x >= topLeft.x) && (point.x <= bottomRight.x) &&
+             (point.y >= topLeft.y) && (point.y <= bottomRight.y);
+   }
+
+   //! Returns true if the box contains the specified rectangle.
+   bool Contains(const RectType<DataType>& rect) const noexcept
+   {
+      return (rect.origin.x >= topLeft.x) &&
+             (rect.origin.x + rect.size.width <= bottomRight.x) &&
+             (rect.origin.y >= topLeft.y) &&
+             (rect.origin.y + rect.size.height <= bottomRight.y);
```

Comment:
Wrongly returns false when the true bottom or right edge of the rectangle coincides with the included boundary.

Fix with `- 1` twice

Reply:
It is possible that a box contains both r1 and r2, but does not contain `Intersect(r1, r2)` if they are disjoint, because you canonicalize the empty rectangle to all zeroes.  That seems weird, not corresponding to set-theoretic identities.

Maybe the function should always return true for empty rectangles, regardless where the top left corner is.